tests/_convert/test_ipynb.py

```diff
@@ -0,0 +1,693 @@
+    ]
+
+
+@pytest.mark.xfail(reason="tricky case not yet supported")
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_duplicate_definitions_with_simple_lambda():
+    sources = [
+        "x = 0",
+        "x = lambda x: x**2",
+    ]
+    result = transform_duplicate_definitions(sources)
+    assert result == [
+        "x = 0",
+        "x_1 = x\nx_1 = lambda x: x**2",
+    ]
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_simple_redefinition() -> None:
+    sources = [
+        "x = 0",
+        "x",
+        "x = 1",
+        "x",
+    ]
+    result = _transform_sources(sources, [{} for _ in sources])
+    assert result == [
+        "x = 0",
+        "x",
+        "x_1 = 1",
+        "x_1",
+    ]
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_duplicate_definitions_with_simple_function():
+    sources = [
+        "x = 0",
+        "x = 1",
+        "def f(x): return x",
+    ]
+    result = _transform_sources(sources, [{} for _ in sources])
+    assert result == [
+        "_x = 0",
+        "_x = 1",
+        "def f(x): return x",
+    ]
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_duplicate_definitions_attrs():
+    sources = [
+        "x = 0",
+        "x",
+        "x = x.apply()",
+    ]
+    result = transform_duplicate_definitions(sources)
+    assert result == [
+        "x = 0",
+        "x",
+        "x_1 = x\nx_1 = x_1.apply()",
+    ]
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_duplicate_definition_shadowed_definition():
+    sources = dd(
+        [
+            "x = 0",
+            "x",
+            """
+            x = 1
+            def f():
+                x = 1;
+            """,
+        ]
+    )
+    result = transform_duplicate_definitions(sources)
+    assert_sources_equal(
+        result,
+        [
+            "x = 0",
+            "x",
+            """
+            x_1 = 1
+            def f():
+                x = 1;
+            """,
+        ],
+    )
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 9), reason="Feature not supported in python 3.8"
+)
+def test_transform_duplicate_definition_kwarg():
+    sources = dd(
+        [
+            "x = 0",
+            "x",
+            """
+            x = 1
+            def f(x=x):
+                return x
+            """,
+        ]
+    )
+    result = transform_duplicate_definitions(sources)
+    assert_sources_equal(
+        result,
+        [
+            "x = 0",
+            "x",
+            """
+            x_1 = 1
+            def f(x=x_1):
+                return x
+            """,
+        ],
+    )
```

Comment:
These are failing, but i think we can fix them by hooking the transformer into the `ScopedVisitor`